markdown/spring-cloud-stream-beyond/spring-cloud-stream-beyond.md

```diff
@@ -0,0 +1,620 @@
+            consumer:
+              queue-access-type: 1 #1 is Exclusive; 0 is Non-Exclusive (and default)
+```
+
+Now if you restart your app you'll see that a durable queue was created. 
+```
+SolaceQueueProvisioner : Subscribing queue scst/wk/exclusive/plain/spring/cloud/stream to topic spring/cloud/stream
+```
+
+And if you look in PubSub+ Manager you'll find that the queue is "Exclusive" which means that the first microservice which connects to it will receive all the messages in the order they are received by the broker. 
+![Exclusive Queue](img/exclusivequeue.webp)
+
+🛠 Go ahead and test it out by sending more messages with the "Try-Me" tab. Now that the queue is durable you can even stop your app, send a bunch of messages and then start it back up to see that they are delivered in order. 
+
+
+### Consumer Groups
+Another popular consuming pattern is using what are called **Consumer Groups**. Consumer groups in Spring Cloud Stream allow you to have multiple consumers sharing the processing of events. 
+To do this you just need to specify a `group` on your input binding. 
+
+To configure this on our app that has the `myConsumer` Consumer function your config would look like this.
+```
+spring:
+  cloud:
+    function:
+      definition: myConsumer
+    stream:
+      bindings:
+        myConsumer-in-0:
+          destination: spring/cloud/stream
+          group: nonexclusive
+```
+
+When using the Solace Binder and specifying a `group` the binder will actually create a Durable Non-Exclusive Queue Endpoint by default. This durable queue will hold messages for your microservices if they get disconnected. 
+Go ahead and run the app, you'll see the following log message that specifies the queue. 
+```
+SolaceQueueProvisioner : Subscribing queue scst/wk/nonexclusive/plain/spring/cloud/stream to topic spring/cloud/stream
+```
+
+And in the PubSub+ Manager you'll see that it created a non-exclusive queue:
+![NonExclusive Queue](img/nonexclusivequeue.webp)
+
+#### Scaling
+So we're using the Consumer Group pattern (which uses a Non-Exclusive Queue!) so we can have multiple consumers share the processing of events. What are the choices for scaling!?
+
+**Option 1 - Multiple Instances of your Microservice**
+The first option is just to start up more instances of your microservice. This is common in a Kubernetes environment where your microservice runs in it's own container. Each instance of the microservice will connect to the same non-exclusive queue on the Solace broker and the broker will spread the messages across all consumers. 
+
+**Option 2 - Concurrent Message Consumption in a Single Microservice**
+The second option is to use the [concurrency consumer property](https://docs.spring.io/spring-cloud-stream/docs/3.1.1/reference/html/spring-cloud-stream.html#_consumer_properties) to enable concurrent message consumption for a particular consumer binding. Under the covers the Solace binder will create a separate flow for 
```

Comment:
> separate flow for every thread?

Reply:
yep! it creates a separate flow for each concurrent consumer which can handle the message on it's own thread. Note that the Solace Session (connection) is shared across all the flows :)

Reply:
jsut read it...fixed the sentence